### PR TITLE
feat: getProposals and getDetails refactor

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/__tests__/Governance.ts [hunk 1]
```diff
@@ -4,18 +4,22 @@ import BigNumber from 'bignumber.js';
 import sinon from 'sinon';
```
Comment (monitz87) [inline src/__tests__/Governance.ts:4]: None of the changes in this file are necessary

File: src/api/entities/Proposal/index.ts [hunk 2]
```diff
@@ -147,25 +147,46 @@ export class Proposal extends Entity<UniqueIdentifiers> {
...
-      module: sectionName,
-      method: methodName,
+      pipId: id,
+      proposer,
+      createdAt,
+      url,
+      description,
+      coolOffEndBlock,
+      endBlock,
+      call: rawProposal ? meshProposalToCall(rawProposal, context) : undefined,
```
Comment (monitz87) [inline src/api/entities/Proposal/index.ts:184]: Why wouldn't `rawProposal` be defined?

File: src/api/entities/Proposal/types.ts [hunk 3]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
...
+
+export interface ProposalDetails {
+  pipId: number;
+  proposer: string;
+  createdAt: number;
+  url: string;
+  description: string;
+  coolOffEndBlock: number;
+  endBlock: number;
+  call?: Call;
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:36]: This should be `transaction: TxTag`

File: src/api/entities/Proposal/types.ts [hunk 4]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
   Ended = 'Ended',
 }
 
-export interface ProposalDetails {
-  state: ProposalState;
-  module: string;
+export interface Call {
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:23]: No need for this interface

File: src/api/entities/Proposal/types.ts (hunk 3 above)
Comment (monitz87) [reply src/api/entities/Proposal/types.ts:36]: Make sure when you're converting that the method and module strings follow the same format we use in TxTags

File: src/api/entities/Proposal/types.ts [hunk 5]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
...
   method: string;
+  module: string;
+}
+
+export interface ProposalDetails {
+  pipId: number;
+  proposer: string;
+  createdAt: number;
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:31]: What is this? A block number? Timestamp?

File: src/api/entities/Proposal/types.ts [hunk 6]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
...
+  url: string;
+  description: string;
+  coolOffEndBlock: number;
+  endBlock: number;
+  call?: Call;
+  lastState: ProposalState;
+  lastStateUpdatedAt: number;
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:38]: Block number or timestamp?

File: src/api/entities/Proposal/types.ts [hunk 7]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
...
-export interface ProposalDetails {
-  state: ProposalState;
-  module: string;
+export interface Call {
   method: string;
+  module: string;
+}
+
+export interface ProposalDetails {
+  pipId: number;
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:29]: No need for this, the `Proposal` entity already has it

File: src/api/entities/Proposal/types.ts [hunk 8]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
...
+  proposer: string;
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:30]: `proposerAddress`

File: src/api/entities/Proposal/types.ts [hunk 9]
```diff
@@ -51,10 +20,25 @@ export enum ProposalStage {
...
+  createdAt: number;
+  url: string;
```
Comment (monitz87) [inline src/api/entities/Proposal/types.ts:32]: `discussionUrl`

File: src/utils/__tests__/index.ts [hunk 10]
```diff
@@ -2419,3 +2420,35 @@ describe('toIdentityWithClaimsArray', () => {
...
+
+  afterEach(() => {
+    dsMockUtils.reset();
+  });
+
+  afterAll(() => {
+    dsMockUtils.cleanup();
+  });
+
+  test('meshProposalToCall should convert a hexa string to a Call object', () => {
```
Comment (monitz87) [inline src/utils/__tests__/index.ts:2437]:
```suggestion
  test('meshProposalToCall should convert a hex string to a Call object', () => {
```

File: src/utils/index.ts [hunk 11]
```diff
@@ -1339,3 +1339,14 @@ export function toIdentityWithClaimsArray(
     ),
   }));
 }
+
+/**
+ * @hidden
+ */
+export function meshProposalToCall(bytes: string, context: Context): Call {
```
Comment (monitz87) [inline src/utils/index.ts:1346]: middlewareProposalToTxTag

File: src/utils/index.ts [hunk 12]
```diff
@@ -1339,3 +1339,14 @@ export function toIdentityWithClaimsArray(
...
+  const call = context.polymeshApi.createType('Proposal', bytes);
+  return {
+    method: call.methodName,
+    module: call.sectionName,
+  };
```
Comment (monitz87) [inline src/utils/index.ts:1351]: You should use destructuring here

File: src/api/entities/Proposal/__tests__/index.ts [hunk 13]
```diff
@@ -8,7 +7,7 @@ import { ProposalStage } from '~/api/entities/Proposal/types';
 import { cancelProposal, editProposal } from '~/api/procedures';
 import { Entity, TransactionQueue } from '~/base';
 import { Context } from '~/context';
-import { eventByIndexedArgs, proposalVotes } from '~/middleware/queries';
+import { eventByIndexedArgs, proposal as queryProposal, proposalVotes } from '~/middleware/queries';
```
Comment (monitz87) [inline src/api/entities/Proposal/__tests__/index.ts:10]: proposalQuery instead of queryProposal

File: src/api/entities/Proposal/index.ts (hunk 2 above)
Comment (shuffledex) [reply src/api/entities/Proposal/index.ts:184]: Proposals that uses `setCode` are returned as null by the harvester

File: src/api/entities/Proposal/types.ts (hunk 5 above)
Comment (shuffledex) [reply src/api/entities/Proposal/types.ts:31]: block number

File: src/api/entities/Proposal/types.ts (hunk 6 above)
Comment (shuffledex) [reply src/api/entities/Proposal/types.ts:38]: block number

File: src/api/entities/Proposal/types.ts (hunk 5 above)
Comment (monitz87) [reply src/api/entities/Proposal/types.ts:26]: Should be BigNumber then

File: src/api/entities/Proposal/types.ts (hunk 6 above)
Comment (monitz87) [reply src/api/entities/Proposal/types.ts:33]: Should be BigNumber then

File: src/utils/index.ts [hunk 14]
```diff
@@ -1355,10 +1355,16 @@ export function toIdentityWithClaimsArray(
...
+  const { sectionName, methodName } = proposal;
+
+  let moduleName;
+  for (const txTagItem in TxTags) {
+    if (txTagItem.toLowerCase() === sectionName.toLowerCase()) {
+      moduleName = txTagItem;
+    }
+  }
+
+  return `${moduleName}.${camelCase(methodName)}` as TxTag;
```
Comment (monitz87) [inline src/utils/index.ts:1369]: Why copy/paste the code from `extrinsicIdentifierToTxTag` when you can just call the function?

Comment (monitz87) [reply src/utils/index.ts:1369]: (Just remember to use .toLowerCase on the original's moduleName)

File: src/Governance.ts [hunk 15]
```diff
@@ -97,14 +103,42 @@ export class Governance {
...
+        totalVotes,
+        totalAyesWeight,
+        totalNaysWeight,
+      }) => {
         const proposal = new Proposal({ pipId }, context);
         return {
           proposal,
-          details: await proposal.getDetails(),
+          details: {
+            pipId,
```
Comment (monitz87) [inline src/Governance.ts:126]: This shouldn't be here

File: src/utils/index.ts [hunk 16]
```diff
@@ -1355,10 +1355,16 @@ export function toIdentityWithClaimsArray(
...
  */
-export function meshProposalToCall(bytes: string, context: Context): Call {
-  const call = context.polymeshApi.createType('Proposal', bytes);
-  return {
-    method: call.methodName,
-    module: call.sectionName,
-  };
+export function middlewareProposalToTxTag(bytes: string, context: Context): TxTag {
+  const proposal = context.polymeshApi.createType('Proposal', bytes);
+  const { sectionName, methodName } = proposal;
```
Comment (monitz87) [inline src/utils/index.ts:1360]:
```suggestion
  const { sectionName, methodName } = context.polymeshApi.createType('Proposal', bytes);
```

File: src/Governance.ts [hunk 17]
```diff
@@ -89,7 +103,44 @@ export class Governance {
...
+            coolOffEndBlock: new BigNumber(coolOffEndBlock),
+            endBlock: new BigNumber(endBlock),
+            transaction: rawProposal ? middlewareProposalToTxTag(rawProposal, context) : null,
+            lastState,
+            lastStateUpdatedAt: new BigNumber(lastStateUpdatedAt),
+            totalVotes: new BigNumber(totalVotes),
+            totalAyesWeight: new BigNumber(totalAyesWeight),
+            totalNaysWeight: new BigNumber(totalNaysWeight),
+          },
+        };
```
Comment (monitz87) [inline src/Governance.ts:139]: This code is too similar to the one in `getDetails`. You should create a util, something like `middlewareProposalToProposalDetails` and use it in both places

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/__tests__/Governance.ts [hunk 18]
```diff
@@ -99,16 +99,39 @@ describe('Governance class', () => {
...
+        ...proposalsQueryResponse,
+        transaction: fakeTransaction,
+        createdAt,
+        coolOffEndBlock,
+        endBlock,
+        lastStateUpdatedAt,
+        totalVotes,
+        totalAyesWeight,
+        totalNaysWeight,
+      });
```
Comment (monitz87) [inline src/__tests__/Governance.ts:134]: Isn't this the same as `fakeResult[0]`?

File: src/utils/index.ts [hunk 19]
```diff
@@ -1363,3 +1364,41 @@ export function middlewareProposalToTxTag(bytes: string, context: Context): TxTa
...
+  } = proposal;
+
+  return {
+    proposerAddress,
+    createdAt: new BigNumber(createdAt),
+    discussionUrl,
+    description,
+    coolOffEndBlock: new BigNumber(coolOffEndBlock),
+    endBlock: new BigNumber(endBlock),
+    transaction: rawProposal ? middlewareProposalToTxTag(rawProposal, context) : null,
```
Comment (monitz87) [inline src/utils/index.ts:1399]: `middlewareProposalToTxTag` should be renamed to `transactionHexToTxTag` to avoid confusion

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Review by monitz87 — APPROVED** [review]

**Pull request merged**